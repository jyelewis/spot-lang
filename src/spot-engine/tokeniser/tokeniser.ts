import { spotKeywords, spotSymbols, SpotToken } from './SpotToken';
import { CodeFileReader } from './CodeFileReader';
import assert from 'node:assert';

export function tokeniseCode(code: string): SpotToken[] {
  const codeReader = new CodeFileReader(code);
  const tokens: SpotToken[] = [];
  while (!codeReader.isEOF) {
    // try to match comments
    if (codeReader.peak('//')) {
      const commentLocation = codeReader.cloneLocation();

      // rest of the line is a comment
      codeReader.consume('//');
      const commentText = codeReader.restOfCurrentLine;
      codeReader.consume(commentText);
      tokens.push({
        type: 'comment',
        location: commentLocation,
        comment: commentText,
      });
    }

    // try to match keyword tokens
    let foundKeyword = false;
    for (const keyword of spotKeywords) {
      if (codeReader.consume(keyword)) {
        foundKeyword = true;
        tokens.push({
          type: 'keyword',
          location: codeReader.cloneLocation(),
          keyword,
        });
        break;
      }
    }

    if (foundKeyword) {
      continue;
    }

    // try to match identifier tokens
    const identifierMatch = codeReader.restOfCurrentLine.match(/^([A-Za-z_][A-Z-a-z0-9_]+)/);
    if (identifierMatch) {
      const identifier = identifierMatch[0];
      codeReader.consume(identifier);
      tokens.push({
        type: 'identifier',
        location: codeReader.cloneLocation(),
        identifier,
      });
      continue;
    }

    let foundSymbol = false;
    for (const symbol of spotSymbols) {
      if (codeReader.consume(symbol)) {
        tokens.push({
          type: 'symbol',
          location: codeReader.cloneLocation(),
          symbol,
        });
        foundSymbol = true;
        break;
      }
    }

    if (foundSymbol) {
      continue;
    }

    // try to read a string literal
    if (codeReader.peak('"')) {
      const startLocation = codeReader.cloneLocation();

      // TODO: revisit with more robust string handling
      const endColumn = codeReader.restOfCurrentLine.substring(1).indexOf('"');

      const literal = codeReader.restOfCurrentLine.substring(1, endColumn + 1);
      const didConsume = codeReader.consume(`"${literal}"`);
      assert(didConsume, 'Should have consumed the string literal');

      tokens.push({
        type: 'string_literal',
        location: startLocation,
        literal,
      });
      continue;
    }

    // try to match a float literal
    const floatMatch = codeReader.restOfCurrentLine.match(/^([0-9]+\.[0-9]+)/);
    if (floatMatch !== null) {
      const floatLocation = codeReader.cloneLocation();
      const floatString = floatMatch[0];
      const floatValue = parseFloat(floatString);
      assert(!isNaN(floatValue), 'Should have parsed a float');

      const didConsume = codeReader.consume(floatString);
      assert(didConsume, 'Should have consumed the float literal');

      tokens.push({
        type: 'float',
        location: floatLocation,
        float: floatValue,
      });
      continue;
    }

    // try to read an integer literal
    const intMatch = codeReader.restOfCurrentLine.match(/^([0-9]+)/);
    if (intMatch !== null) {
      const intLocation = codeReader.cloneLocation();
      const intString = intMatch[0];
      const intValue = parseInt(intString, 10);
      assert(!isNaN(intValue), 'Should have parsed an integer');

      const didConsume = codeReader.consume(intString);
      assert(didConsume, 'Should have consumed the integer literal');

      tokens.push({
        type: 'int',
        location: intLocation,
        int: intValue,
      });
      continue;
    }

    throw new Error(
      'Unrecognised token at ' +
        JSON.stringify(codeReader.cloneLocation()) +
        ': "' +
        codeReader.restOfCurrentLine.slice(0, 10) +
        '"' +
        (codeReader.restOfCurrentLine.length > 10 ? '...' : '')
    );
  }

  return tokens;
}
