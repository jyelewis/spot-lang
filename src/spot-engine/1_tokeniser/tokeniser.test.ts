import { tokeniseCode } from './tokeniser';
import path from 'node:path';
import fs from 'node:fs/promises';

const examplesDir = `${__dirname}/../../../docs/code_examples`;

describe('Tokeniser', () => {
  describe('code_examples folder', () => {
    const tokeniseExample = async (exampleName: string) => {
      const codeFilePath = path.join(examplesDir, `${exampleName}.spot`);
      const code = await fs.readFile(codeFilePath, 'utf-8');

      return await tokeniseCode(code);
    };

    it(`Tokenises '0_print_hello.spot`, () =>
      expect(tokeniseExample('0_print_hello')).resolves.toMatchSnapshot());

    it(`Tokenises '1_variables.spot`, () =>
      expect(tokeniseExample('1_variables')).resolves.toMatchSnapshot());

    it(`Tokenises '2_addition.spot`, () =>
      expect(tokeniseExample('2_addition')).resolves.toMatchSnapshot());

    it(`Tokenises '3_ifs.spot`, () => expect(tokeniseExample('3_ifs')).resolves.toMatchSnapshot());
  });

  describe('Specific case', () => {
    it('Reads hello world example', () => {
      const code = `
    
func main() {
  print("Hello, World!")
}
`;
      const tokens = tokeniseCode(code);
      expect(tokens).toMatchSnapshot();
    });

    it('Variable definitions', () => {
      const code = `
    
func main() {
  let name = "John Doe"
  print("Hello, World!")
}
`;
      const tokens = tokeniseCode(code);
      expect(tokens).toMatchSnapshot();
    });

    it('Comments', () => {
      const code = `
    
func main() {
  // this is a comment
}
`;
      const tokens = tokeniseCode(code);
      expect(
        tokens.find((x) => x.type === 'comment' && x.comment === 'this is a comment')
      ).toBeDefined();
      expect(tokens).toMatchSnapshot();
    });

    it('Integers', () => {
      const code = `
func main() {
  let num = 123
}
`;
      const tokens = tokeniseCode(code);
      expect(tokens.find((x) => x.type === 'int' && x.int === 123)).toBeDefined();
      expect(tokens).toMatchSnapshot();
    });

    it('Floats', () => {
      const code = `
func main() {
  let num = 123.456
}
`;
      const tokens = tokeniseCode(code);
      expect(tokens.find((x) => x.type === 'float' && x.float === 123.456)).toBeDefined();
      expect(tokens).toMatchSnapshot();
    });

    it('Int addition', () => {
      const code = `
func main() {
  let num = (123 + 456) - 1
}
`;
      const tokens = tokeniseCode(code);
      expect(tokens).toMatchSnapshot();
    });

    it('String addition', () => {
      const code = `
func main() {
  let name = "John" + " Doe"
}
`;
      const tokens = tokeniseCode(code);
      expect(tokens).toMatchSnapshot();
    });

    it('Variable addition', () => {
      const code = `
func main() {
  let first_name = "John"
  let last_name = "Doe"
  
  let name = first_name + " " + last_name
}
`;
      const tokens = tokeniseCode(code);
      expect(tokens).toMatchSnapshot();
    });

    // TODO: parsed string template support
  });
});
