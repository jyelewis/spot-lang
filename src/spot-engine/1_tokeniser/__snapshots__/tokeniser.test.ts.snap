// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tokeniser Specific case Comments 1`] = `
[
  {
    "keyword": "func",
    "location": {
      "column": 1,
      "line": 3,
    },
    "type": "keyword",
  },
  {
    "identifier": "main",
    "location": {
      "column": 6,
      "line": 3,
    },
    "type": "identifier",
  },
  {
    "location": {
      "column": 10,
      "line": 3,
    },
    "symbol": "(",
    "type": "symbol",
  },
  {
    "location": {
      "column": 11,
      "line": 3,
    },
    "symbol": ")",
    "type": "symbol",
  },
  {
    "location": {
      "column": 13,
      "line": 3,
    },
    "symbol": "{",
    "type": "symbol",
  },
  {
    "comment": "this is a comment",
    "location": {
      "column": 3,
      "line": 4,
    },
    "type": "comment",
  },
  {
    "location": {
      "column": 1,
      "line": 5,
    },
    "symbol": "}",
    "type": "symbol",
  },
]
`;

exports[`Tokeniser Specific case Floats 1`] = `
[
  {
    "keyword": "func",
    "location": {
      "column": 1,
      "line": 2,
    },
    "type": "keyword",
  },
  {
    "identifier": "main",
    "location": {
      "column": 6,
      "line": 2,
    },
    "type": "identifier",
  },
  {
    "location": {
      "column": 10,
      "line": 2,
    },
    "symbol": "(",
    "type": "symbol",
  },
  {
    "location": {
      "column": 11,
      "line": 2,
    },
    "symbol": ")",
    "type": "symbol",
  },
  {
    "location": {
      "column": 13,
      "line": 2,
    },
    "symbol": "{",
    "type": "symbol",
  },
  {
    "keyword": "let",
    "location": {
      "column": 3,
      "line": 3,
    },
    "type": "keyword",
  },
  {
    "identifier": "num",
    "location": {
      "column": 7,
      "line": 3,
    },
    "type": "identifier",
  },
  {
    "location": {
      "column": 11,
      "line": 3,
    },
    "symbol": "=",
    "type": "symbol",
  },
  {
    "float": 123.456,
    "location": {
      "column": 13,
      "line": 3,
    },
    "type": "float",
  },
  {
    "location": {
      "column": 1,
      "line": 4,
    },
    "symbol": "}",
    "type": "symbol",
  },
]
`;

exports[`Tokeniser Specific case Int addition 1`] = `
[
  {
    "keyword": "func",
    "location": {
      "column": 1,
      "line": 2,
    },
    "type": "keyword",
  },
  {
    "identifier": "main",
    "location": {
      "column": 6,
      "line": 2,
    },
    "type": "identifier",
  },
  {
    "location": {
      "column": 10,
      "line": 2,
    },
    "symbol": "(",
    "type": "symbol",
  },
  {
    "location": {
      "column": 11,
      "line": 2,
    },
    "symbol": ")",
    "type": "symbol",
  },
  {
    "location": {
      "column": 13,
      "line": 2,
    },
    "symbol": "{",
    "type": "symbol",
  },
  {
    "keyword": "let",
    "location": {
      "column": 3,
      "line": 3,
    },
    "type": "keyword",
  },
  {
    "identifier": "num",
    "location": {
      "column": 7,
      "line": 3,
    },
    "type": "identifier",
  },
  {
    "location": {
      "column": 11,
      "line": 3,
    },
    "symbol": "=",
    "type": "symbol",
  },
  {
    "location": {
      "column": 13,
      "line": 3,
    },
    "symbol": "(",
    "type": "symbol",
  },
  {
    "int": 123,
    "location": {
      "column": 14,
      "line": 3,
    },
    "type": "int",
  },
  {
    "location": {
      "column": 18,
      "line": 3,
    },
    "symbol": "+",
    "type": "symbol",
  },
  {
    "int": 456,
    "location": {
      "column": 20,
      "line": 3,
    },
    "type": "int",
  },
  {
    "location": {
      "column": 23,
      "line": 3,
    },
    "symbol": ")",
    "type": "symbol",
  },
  {
    "location": {
      "column": 25,
      "line": 3,
    },
    "symbol": "-",
    "type": "symbol",
  },
  {
    "int": 1,
    "location": {
      "column": 27,
      "line": 3,
    },
    "type": "int",
  },
  {
    "location": {
      "column": 1,
      "line": 4,
    },
    "symbol": "}",
    "type": "symbol",
  },
]
`;

exports[`Tokeniser Specific case Integers 1`] = `
[
  {
    "keyword": "func",
    "location": {
      "column": 1,
      "line": 2,
    },
    "type": "keyword",
  },
  {
    "identifier": "main",
    "location": {
      "column": 6,
      "line": 2,
    },
    "type": "identifier",
  },
  {
    "location": {
      "column": 10,
      "line": 2,
    },
    "symbol": "(",
    "type": "symbol",
  },
  {
    "location": {
      "column": 11,
      "line": 2,
    },
    "symbol": ")",
    "type": "symbol",
  },
  {
    "location": {
      "column": 13,
      "line": 2,
    },
    "symbol": "{",
    "type": "symbol",
  },
  {
    "keyword": "let",
    "location": {
      "column": 3,
      "line": 3,
    },
    "type": "keyword",
  },
  {
    "identifier": "num",
    "location": {
      "column": 7,
      "line": 3,
    },
    "type": "identifier",
  },
  {
    "location": {
      "column": 11,
      "line": 3,
    },
    "symbol": "=",
    "type": "symbol",
  },
  {
    "int": 123,
    "location": {
      "column": 13,
      "line": 3,
    },
    "type": "int",
  },
  {
    "location": {
      "column": 1,
      "line": 4,
    },
    "symbol": "}",
    "type": "symbol",
  },
]
`;

exports[`Tokeniser Specific case Reads hello world example 1`] = `
[
  {
    "keyword": "func",
    "location": {
      "column": 1,
      "line": 3,
    },
    "type": "keyword",
  },
  {
    "identifier": "main",
    "location": {
      "column": 6,
      "line": 3,
    },
    "type": "identifier",
  },
  {
    "location": {
      "column": 10,
      "line": 3,
    },
    "symbol": "(",
    "type": "symbol",
  },
  {
    "location": {
      "column": 11,
      "line": 3,
    },
    "symbol": ")",
    "type": "symbol",
  },
  {
    "location": {
      "column": 13,
      "line": 3,
    },
    "symbol": "{",
    "type": "symbol",
  },
  {
    "identifier": "print",
    "location": {
      "column": 3,
      "line": 4,
    },
    "type": "identifier",
  },
  {
    "location": {
      "column": 8,
      "line": 4,
    },
    "symbol": "(",
    "type": "symbol",
  },
  {
    "literal": "Hello, World!",
    "location": {
      "column": 9,
      "line": 4,
    },
    "type": "string_literal",
  },
  {
    "location": {
      "column": 24,
      "line": 4,
    },
    "symbol": ")",
    "type": "symbol",
  },
  {
    "location": {
      "column": 1,
      "line": 5,
    },
    "symbol": "}",
    "type": "symbol",
  },
]
`;

exports[`Tokeniser Specific case String addition 1`] = `
[
  {
    "keyword": "func",
    "location": {
      "column": 1,
      "line": 2,
    },
    "type": "keyword",
  },
  {
    "identifier": "main",
    "location": {
      "column": 6,
      "line": 2,
    },
    "type": "identifier",
  },
  {
    "location": {
      "column": 10,
      "line": 2,
    },
    "symbol": "(",
    "type": "symbol",
  },
  {
    "location": {
      "column": 11,
      "line": 2,
    },
    "symbol": ")",
    "type": "symbol",
  },
  {
    "location": {
      "column": 13,
      "line": 2,
    },
    "symbol": "{",
    "type": "symbol",
  },
  {
    "keyword": "let",
    "location": {
      "column": 3,
      "line": 3,
    },
    "type": "keyword",
  },
  {
    "identifier": "name",
    "location": {
      "column": 7,
      "line": 3,
    },
    "type": "identifier",
  },
  {
    "location": {
      "column": 12,
      "line": 3,
    },
    "symbol": "=",
    "type": "symbol",
  },
  {
    "literal": "John",
    "location": {
      "column": 14,
      "line": 3,
    },
    "type": "string_literal",
  },
  {
    "location": {
      "column": 21,
      "line": 3,
    },
    "symbol": "+",
    "type": "symbol",
  },
  {
    "literal": " Doe",
    "location": {
      "column": 23,
      "line": 3,
    },
    "type": "string_literal",
  },
  {
    "location": {
      "column": 1,
      "line": 4,
    },
    "symbol": "}",
    "type": "symbol",
  },
]
`;

exports[`Tokeniser Specific case Variable addition 1`] = `
[
  {
    "keyword": "func",
    "location": {
      "column": 1,
      "line": 2,
    },
    "type": "keyword",
  },
  {
    "identifier": "main",
    "location": {
      "column": 6,
      "line": 2,
    },
    "type": "identifier",
  },
  {
    "location": {
      "column": 10,
      "line": 2,
    },
    "symbol": "(",
    "type": "symbol",
  },
  {
    "location": {
      "column": 11,
      "line": 2,
    },
    "symbol": ")",
    "type": "symbol",
  },
  {
    "location": {
      "column": 13,
      "line": 2,
    },
    "symbol": "{",
    "type": "symbol",
  },
  {
    "keyword": "let",
    "location": {
      "column": 3,
      "line": 3,
    },
    "type": "keyword",
  },
  {
    "identifier": "first_name",
    "location": {
      "column": 7,
      "line": 3,
    },
    "type": "identifier",
  },
  {
    "location": {
      "column": 18,
      "line": 3,
    },
    "symbol": "=",
    "type": "symbol",
  },
  {
    "literal": "John",
    "location": {
      "column": 20,
      "line": 3,
    },
    "type": "string_literal",
  },
  {
    "keyword": "let",
    "location": {
      "column": 3,
      "line": 4,
    },
    "type": "keyword",
  },
  {
    "identifier": "last_name",
    "location": {
      "column": 7,
      "line": 4,
    },
    "type": "identifier",
  },
  {
    "location": {
      "column": 17,
      "line": 4,
    },
    "symbol": "=",
    "type": "symbol",
  },
  {
    "literal": "Doe",
    "location": {
      "column": 19,
      "line": 4,
    },
    "type": "string_literal",
  },
  {
    "keyword": "let",
    "location": {
      "column": 3,
      "line": 6,
    },
    "type": "keyword",
  },
  {
    "identifier": "name",
    "location": {
      "column": 7,
      "line": 6,
    },
    "type": "identifier",
  },
  {
    "location": {
      "column": 12,
      "line": 6,
    },
    "symbol": "=",
    "type": "symbol",
  },
  {
    "identifier": "first_name",
    "location": {
      "column": 14,
      "line": 6,
    },
    "type": "identifier",
  },
  {
    "location": {
      "column": 25,
      "line": 6,
    },
    "symbol": "+",
    "type": "symbol",
  },
  {
    "literal": " ",
    "location": {
      "column": 27,
      "line": 6,
    },
    "type": "string_literal",
  },
  {
    "location": {
      "column": 31,
      "line": 6,
    },
    "symbol": "+",
    "type": "symbol",
  },
  {
    "identifier": "last_name",
    "location": {
      "column": 33,
      "line": 6,
    },
    "type": "identifier",
  },
  {
    "location": {
      "column": 1,
      "line": 7,
    },
    "symbol": "}",
    "type": "symbol",
  },
]
`;

exports[`Tokeniser Specific case Variable definitions 1`] = `
[
  {
    "keyword": "func",
    "location": {
      "column": 1,
      "line": 3,
    },
    "type": "keyword",
  },
  {
    "identifier": "main",
    "location": {
      "column": 6,
      "line": 3,
    },
    "type": "identifier",
  },
  {
    "location": {
      "column": 10,
      "line": 3,
    },
    "symbol": "(",
    "type": "symbol",
  },
  {
    "location": {
      "column": 11,
      "line": 3,
    },
    "symbol": ")",
    "type": "symbol",
  },
  {
    "location": {
      "column": 13,
      "line": 3,
    },
    "symbol": "{",
    "type": "symbol",
  },
  {
    "keyword": "let",
    "location": {
      "column": 3,
      "line": 4,
    },
    "type": "keyword",
  },
  {
    "identifier": "name",
    "location": {
      "column": 7,
      "line": 4,
    },
    "type": "identifier",
  },
  {
    "location": {
      "column": 12,
      "line": 4,
    },
    "symbol": "=",
    "type": "symbol",
  },
  {
    "literal": "John Doe",
    "location": {
      "column": 14,
      "line": 4,
    },
    "type": "string_literal",
  },
  {
    "identifier": "print",
    "location": {
      "column": 3,
      "line": 5,
    },
    "type": "identifier",
  },
  {
    "location": {
      "column": 8,
      "line": 5,
    },
    "symbol": "(",
    "type": "symbol",
  },
  {
    "literal": "Hello, World!",
    "location": {
      "column": 9,
      "line": 5,
    },
    "type": "string_literal",
  },
  {
    "location": {
      "column": 24,
      "line": 5,
    },
    "symbol": ")",
    "type": "symbol",
  },
  {
    "location": {
      "column": 1,
      "line": 6,
    },
    "symbol": "}",
    "type": "symbol",
  },
]
`;

exports[`Tokeniser code_examples folder Tokenises '0_print_hello.spot 1`] = `
[
  {
    "keyword": "func",
    "location": {
      "column": 1,
      "line": 2,
    },
    "type": "keyword",
  },
  {
    "identifier": "main",
    "location": {
      "column": 6,
      "line": 2,
    },
    "type": "identifier",
  },
  {
    "location": {
      "column": 10,
      "line": 2,
    },
    "symbol": "(",
    "type": "symbol",
  },
  {
    "location": {
      "column": 11,
      "line": 2,
    },
    "symbol": ")",
    "type": "symbol",
  },
  {
    "location": {
      "column": 13,
      "line": 2,
    },
    "symbol": "{",
    "type": "symbol",
  },
  {
    "identifier": "print",
    "location": {
      "column": 5,
      "line": 3,
    },
    "type": "identifier",
  },
  {
    "location": {
      "column": 10,
      "line": 3,
    },
    "symbol": "(",
    "type": "symbol",
  },
  {
    "literal": "Hello, world!",
    "location": {
      "column": 11,
      "line": 3,
    },
    "type": "string_literal",
  },
  {
    "location": {
      "column": 26,
      "line": 3,
    },
    "symbol": ")",
    "type": "symbol",
  },
  {
    "location": {
      "column": 1,
      "line": 4,
    },
    "symbol": "}",
    "type": "symbol",
  },
]
`;

exports[`Tokeniser code_examples folder Tokenises '1_variables.spot 1`] = `
[
  {
    "keyword": "func",
    "location": {
      "column": 1,
      "line": 2,
    },
    "type": "keyword",
  },
  {
    "identifier": "main",
    "location": {
      "column": 6,
      "line": 2,
    },
    "type": "identifier",
  },
  {
    "location": {
      "column": 10,
      "line": 2,
    },
    "symbol": "(",
    "type": "symbol",
  },
  {
    "location": {
      "column": 11,
      "line": 2,
    },
    "symbol": ")",
    "type": "symbol",
  },
  {
    "location": {
      "column": 13,
      "line": 2,
    },
    "symbol": "{",
    "type": "symbol",
  },
  {
    "keyword": "let",
    "location": {
      "column": 5,
      "line": 3,
    },
    "type": "keyword",
  },
  {
    "identifier": "name",
    "location": {
      "column": 9,
      "line": 3,
    },
    "type": "identifier",
  },
  {
    "location": {
      "column": 14,
      "line": 3,
    },
    "symbol": "=",
    "type": "symbol",
  },
  {
    "literal": "John Doe",
    "location": {
      "column": 16,
      "line": 3,
    },
    "type": "string_literal",
  },
  {
    "keyword": "let",
    "location": {
      "column": 5,
      "line": 4,
    },
    "type": "keyword",
  },
  {
    "identifier": "age",
    "location": {
      "column": 9,
      "line": 4,
    },
    "type": "identifier",
  },
  {
    "location": {
      "column": 13,
      "line": 4,
    },
    "symbol": "=",
    "type": "symbol",
  },
  {
    "int": 36,
    "location": {
      "column": 15,
      "line": 4,
    },
    "type": "int",
  },
  {
    "identifier": "print",
    "location": {
      "column": 5,
      "line": 6,
    },
    "type": "identifier",
  },
  {
    "location": {
      "column": 10,
      "line": 6,
    },
    "symbol": "(",
    "type": "symbol",
  },
  {
    "literal": "Hello, my name is {name} and I am {age} years old.",
    "location": {
      "column": 11,
      "line": 6,
    },
    "type": "string_literal",
  },
  {
    "location": {
      "column": 63,
      "line": 6,
    },
    "symbol": ")",
    "type": "symbol",
  },
  {
    "location": {
      "column": 1,
      "line": 7,
    },
    "symbol": "}",
    "type": "symbol",
  },
]
`;

exports[`Tokeniser code_examples folder Tokenises '2_addition.spot 1`] = `
[
  {
    "keyword": "func",
    "location": {
      "column": 1,
      "line": 2,
    },
    "type": "keyword",
  },
  {
    "identifier": "main",
    "location": {
      "column": 6,
      "line": 2,
    },
    "type": "identifier",
  },
  {
    "location": {
      "column": 10,
      "line": 2,
    },
    "symbol": "(",
    "type": "symbol",
  },
  {
    "location": {
      "column": 11,
      "line": 2,
    },
    "symbol": ")",
    "type": "symbol",
  },
  {
    "location": {
      "column": 13,
      "line": 2,
    },
    "symbol": "{",
    "type": "symbol",
  },
  {
    "keyword": "let",
    "location": {
      "column": 5,
      "line": 3,
    },
    "type": "keyword",
  },
  {
    "identifier": "sum",
    "location": {
      "column": 9,
      "line": 3,
    },
    "type": "identifier",
  },
  {
    "location": {
      "column": 13,
      "line": 3,
    },
    "symbol": "=",
    "type": "symbol",
  },
  {
    "int": 2,
    "location": {
      "column": 15,
      "line": 3,
    },
    "type": "int",
  },
  {
    "location": {
      "column": 17,
      "line": 3,
    },
    "symbol": "+",
    "type": "symbol",
  },
  {
    "int": 3,
    "location": {
      "column": 19,
      "line": 3,
    },
    "type": "int",
  },
  {
    "identifier": "print",
    "location": {
      "column": 5,
      "line": 4,
    },
    "type": "identifier",
  },
  {
    "location": {
      "column": 10,
      "line": 4,
    },
    "symbol": "(",
    "type": "symbol",
  },
  {
    "literal": "The sum of 2 and 3 is {sum}",
    "location": {
      "column": 11,
      "line": 4,
    },
    "type": "string_literal",
  },
  {
    "location": {
      "column": 40,
      "line": 4,
    },
    "symbol": ")",
    "type": "symbol",
  },
  {
    "location": {
      "column": 1,
      "line": 5,
    },
    "symbol": "}",
    "type": "symbol",
  },
]
`;

exports[`Tokeniser code_examples folder Tokenises '3_ifs.spot 1`] = `
[
  {
    "keyword": "func",
    "location": {
      "column": 1,
      "line": 2,
    },
    "type": "keyword",
  },
  {
    "identifier": "main",
    "location": {
      "column": 6,
      "line": 2,
    },
    "type": "identifier",
  },
  {
    "location": {
      "column": 10,
      "line": 2,
    },
    "symbol": "(",
    "type": "symbol",
  },
  {
    "location": {
      "column": 11,
      "line": 2,
    },
    "symbol": ")",
    "type": "symbol",
  },
  {
    "location": {
      "column": 13,
      "line": 2,
    },
    "symbol": "{",
    "type": "symbol",
  },
  {
    "comment": "TODO: what is a boolean but an enum with two cases?",
    "location": {
      "column": 5,
      "line": 3,
    },
    "type": "comment",
  },
  {
    "keyword": "let",
    "location": {
      "column": 5,
      "line": 4,
    },
    "type": "keyword",
  },
  {
    "identifier": "isAuthorized",
    "location": {
      "column": 9,
      "line": 4,
    },
    "type": "identifier",
  },
  {
    "location": {
      "column": 22,
      "line": 4,
    },
    "symbol": "=",
    "type": "symbol",
  },
  {
    "identifier": "True",
    "location": {
      "column": 24,
      "line": 4,
    },
    "type": "identifier",
  },
  {
    "keyword": "if",
    "location": {
      "column": 5,
      "line": 6,
    },
    "type": "keyword",
  },
  {
    "identifier": "isAuthorized",
    "location": {
      "column": 8,
      "line": 6,
    },
    "type": "identifier",
  },
  {
    "location": {
      "column": 21,
      "line": 6,
    },
    "symbol": "{",
    "type": "symbol",
  },
  {
    "identifier": "print",
    "location": {
      "column": 9,
      "line": 7,
    },
    "type": "identifier",
  },
  {
    "location": {
      "column": 14,
      "line": 7,
    },
    "symbol": "(",
    "type": "symbol",
  },
  {
    "literal": "Access granted.",
    "location": {
      "column": 15,
      "line": 7,
    },
    "type": "string_literal",
  },
  {
    "location": {
      "column": 32,
      "line": 7,
    },
    "symbol": ")",
    "type": "symbol",
  },
  {
    "location": {
      "column": 5,
      "line": 8,
    },
    "symbol": "}",
    "type": "symbol",
  },
  {
    "keyword": "else",
    "location": {
      "column": 7,
      "line": 8,
    },
    "type": "keyword",
  },
  {
    "location": {
      "column": 12,
      "line": 8,
    },
    "symbol": "{",
    "type": "symbol",
  },
  {
    "identifier": "print",
    "location": {
      "column": 9,
      "line": 9,
    },
    "type": "identifier",
  },
  {
    "location": {
      "column": 14,
      "line": 9,
    },
    "symbol": "(",
    "type": "symbol",
  },
  {
    "literal": "Access denied.",
    "location": {
      "column": 15,
      "line": 9,
    },
    "type": "string_literal",
  },
  {
    "location": {
      "column": 31,
      "line": 9,
    },
    "symbol": ")",
    "type": "symbol",
  },
  {
    "location": {
      "column": 5,
      "line": 10,
    },
    "symbol": "}",
    "type": "symbol",
  },
  {
    "keyword": "if",
    "location": {
      "column": 5,
      "line": 12,
    },
    "type": "keyword",
  },
  {
    "location": {
      "column": 8,
      "line": 12,
    },
    "symbol": "!",
    "type": "symbol",
  },
  {
    "identifier": "isAuthorized",
    "location": {
      "column": 9,
      "line": 12,
    },
    "type": "identifier",
  },
  {
    "location": {
      "column": 22,
      "line": 12,
    },
    "symbol": "{",
    "type": "symbol",
  },
  {
    "identifier": "print",
    "location": {
      "column": 9,
      "line": 13,
    },
    "type": "identifier",
  },
  {
    "location": {
      "column": 14,
      "line": 13,
    },
    "symbol": "(",
    "type": "symbol",
  },
  {
    "literal": "isAuthorized False",
    "location": {
      "column": 15,
      "line": 13,
    },
    "type": "string_literal",
  },
  {
    "location": {
      "column": 35,
      "line": 13,
    },
    "symbol": ")",
    "type": "symbol",
  },
  {
    "location": {
      "column": 5,
      "line": 14,
    },
    "symbol": "}",
    "type": "symbol",
  },
  {
    "keyword": "else",
    "location": {
      "column": 7,
      "line": 14,
    },
    "type": "keyword",
  },
  {
    "location": {
      "column": 12,
      "line": 14,
    },
    "symbol": "{",
    "type": "symbol",
  },
  {
    "identifier": "print",
    "location": {
      "column": 9,
      "line": 15,
    },
    "type": "identifier",
  },
  {
    "location": {
      "column": 14,
      "line": 15,
    },
    "symbol": "(",
    "type": "symbol",
  },
  {
    "literal": "isAuthorized True",
    "location": {
      "column": 15,
      "line": 15,
    },
    "type": "string_literal",
  },
  {
    "location": {
      "column": 34,
      "line": 15,
    },
    "symbol": ")",
    "type": "symbol",
  },
  {
    "location": {
      "column": 5,
      "line": 16,
    },
    "symbol": "}",
    "type": "symbol",
  },
  {
    "location": {
      "column": 1,
      "line": 17,
    },
    "symbol": "}",
    "type": "symbol",
  },
]
`;
